### PR TITLE
Add support for IPv6 NFS mounts

### PR DESCRIPTION
**Pull request by MyUmmaGumma**
Fix for TITUS-5627

Lookup for valid v4 and v6 and then do getaddrinfo and parse results

### CODE REVIEW

File: mount/titus-mount-nfs.c [hunk 1]
```diff
@@ -29,52 +29,66 @@ char nfs4[] = "nfs4";
...
-		fprintf(stderr, "titus-mount-nfs: can't get address for %s: %s\n",
-			hostname, hstrerror(err_ret));
-		return -1;
-	}
-	if (hp->h_length > (int)sizeof(struct in_addr)) {
-		fprintf(stderr, "titus-mount-nfs: got bad hp->h_length");
-		return -1;
-	}
-	memcpy(&addr->sin_addr, hp->h_addr, hp->h_length);
+static int add_addr_options(const char *hostname, char *final_options)
```
Comment (sargun) [inline mount/titus-mount-nfs.c:34]: Any reason this isn't a void?

File: mount/titus-mount-nfs.c [hunk 2]
```diff
@@ -29,52 +29,66 @@ char nfs4[] = "nfs4";
...
+ * (rely on /etc/gai.conf)
+ */
+static int hostname_to_ip_option(const char *hostname, char *final_options)
 {
-	/* For NFS, we must do the dns resolution *here* while we are inside net ns */
-	struct sockaddr_in server_addr;
-	char *ip_string;
+	struct addrinfo *result, *iter;
+	int error;
+	unsigned char buf[sizeof(struct in6_addr)];
```
Comment (sargun) [inline mount/titus-mount-nfs.c:49]: Shouldn't this be buf[INET6_ADDRSTRLEN]?

File: mount/titus-mount-nfs.c [hunk 3]
```diff
@@ -308,7 +322,11 @@ int main(int argc, char *argv[])
 	/* Now we can switch net/mount namespaces so we can lookup the ip and eventually mount */
 	switch_namespaces(pidfd);
 	fprintf(stderr, "titus-mount-nfs: user-inputed options: %s\n", options);
-	compose_final_options(nfs_mount_hostname, final_options);
+	res = hostname_to_ip_option(nfs_mount_hostname, final_options);
+	if (res != 0) {
+		fprintf(stderr, "Error resolving hostname to IP address on mount");
```
Comment (sargun) [inline mount/titus-mount-nfs.c:327]: Add \n at the end.

File: mount/titus-mount-nfs.c [hunk 4]
```diff
@@ -29,52 +29,66 @@ char nfs4[] = "nfs4";
...
+	} 
+
+	error = inet_pton(AF_INET6, hostname, buf);
+	if (error == 1) {
+		fprintf(stderr, "Valid IPv6 address %s - No DNS needed", hostname);
+		add_addr_options(hostname, final_options);
+		return 0;
+	} 
+
+	fprintf(stderr, "Now trying to do getaddrinfo %s: ", gai_strerror(error));
```
Comment (sargun) [inline mount/titus-mount-nfs.c:65]: I think that this should be:

`fprintf(stderr, "Now trying to do getaddrinfo: %s\n", hostname)`

File: mount/titus-mount-nfs.c [hunk 5]
```diff
@@ -29,52 +29,66 @@ char nfs4[] = "nfs4";
...
+	error = getaddrinfo(hostname, NULL, NULL, &result);
+	if (error) {
+		fprintf(stderr, "Error trying to resolve %s: ", gai_strerror(error));
```
Comment (sargun) [inline mount/titus-mount-nfs.c:68]: fprintf(stderr, "Error trying to resolve %s: %s", hostname, gai_strerror(error));

File: executor/runtime/docker/docker.go [hunk 6]
```diff
@@ -1860,7 +1860,7 @@ func (r *DockerRuntime) setupEFSMounts(parentCtx context.Context, c runtimeTypes
 		mountOptions := append(
 			baseMountOptions,
 			fmt.Sprintf("fsc=%s", c.TaskID()),
-			fmt.Sprintf("source=%s:%s", nfs.Server, nfs.ServerPath),
+			fmt.Sprintf("source=[%s]:%s", nfs.Server, nfs.ServerPath),
```
Comment (solarkennedy) [inline executor/runtime/docker/docker.go:1863]: Is putting brackets on the outside here safe to do, safe to do, even if it is a regular hostname or ipv4?

File: mount/titus-mount-nfs.c (hunk 1 above)
Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c]: Fixed.

File: mount/titus-mount-nfs.c (hunk 2 above)
Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c]: Fixed.

File: mount/titus-mount-nfs.c (hunk 3 above)
Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c]: Fixed.

File: mount/titus-mount-nfs.c (hunk 4 above)
Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c]: removed.

File: mount/titus-mount-nfs.c (hunk 5 above)
Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c]: Fixed.

File: executor/runtime/docker/docker.go (hunk 6 above)
Comment (MyUmmaGumma) [reply executor/runtime/docker/docker.go:1863]: Yes, I tested the IPv4 and seemed to work fine. The unit tests tested against the EFS DNS endpoint which also worked.

File: mount/titus-mount-nfs.c [hunk 7]
```diff
@@ -29,52 +29,64 @@ char nfs4[] = "nfs4";
...
+		return -1;
+	} 
+
+	for (iter = result; iter != NULL; iter = iter->ai_next) {
+		char ipaddress[IP_ADDRESS_LEN];
+		void *ptr_to_ip;
+		inet_ntop(iter->ai_family, iter->ai_addr->sa_data, ipaddress, IP_ADDRESS_LEN);
+		switch (iter->ai_family) {
+			case AF_INET:
+				ptr_to_ip = &((struct sockaddr_in *)iter->ai_addr)->sin_addr;
```
Comment (solarkennedy) [inline mount/titus-mount-nfs.c:76]: Is this safe? What if the container doesn't have a v6 address?

Comment (MyUmmaGumma) [reply mount/titus-mount-nfs.c:76]: Code wise its safe but as you pointed out for a Ipv4Only container, this code would prefer resolving to a v6 address and the connection will run into problems. That scenario will require thought where we might want to retry mounting onto the IPv4 resolved address were the Ipv6 mount fail due to network unreachability.

**Comment by sargun** [comment]
Don't all your fprintf's need a \n at the end?

**Review by sargun — APPROVED** [review]

**Pull request merged**